File: Airwallex/Card/AWXUIContext+Card.m
//
//  AWXUIContext+Card.m
//  AirWallexPaymentSDK
//
//  Created by Weiping Li on 2024/11/27.
//

#import "AWXCardViewController.h"
#import "AWXCardViewModel.h"
#import "AWXUIContext+Card.h"

@implementation AWXUIContext (Card)

/// Present the card payment flow.
- (void)presentCardPaymentFlowFrom:(UIViewController *)hostViewController cardSchemes:(NSArray<AWXCardBrand> *)cardSchemes {
    AWXCardViewController *controller = [[AWXCardViewController alloc] initWithNibName:nil bundle:nil];
    controller.session = self.session;

    NSMutableArray<AWXCardScheme *> *cardSchemesArray = [NSMutableArray array];
    for (AWXCardBrand scheme in cardSchemes) {
        AWXCardScheme *cardScheme = [[AWXCardScheme alloc] init];
        cardScheme.name = scheme;
        [cardSchemesArray addObject:cardScheme];
    }

    AWXCardViewModel *viewModel = [[AWXCardViewModel alloc] initWithSession:self.session supportedCardSchemes:cardSchemesArray launchDirectly:YES];
    controller.viewModel = viewModel;

    UINavigationController *navigationController = [[UINavigationController alloc] initWithRootViewController:controller];
    navigationController.modalInPresentation = YES;

    [hostViewController presentViewController:navigationController animated:YES completion:nil];
}

/// Push the card payment flow.
- (void)pushCardPaymentFlowFrom:(UIViewController *)hostViewController cardSchemes:(NSArray<AWXCardBrand> *)cardSchemes {
    UINavigationController *navigationController = nil;

    if ([hostViewController isKindOfClass:[UINavigationController class]]) {
        navigationController = (UINavigationController *)hostViewController;
    } else {
        navigationController = hostViewController.navigationController;
    }

    if (!navigationController) {
        @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                       reason:@"The hostViewController is not a navigation controller, or is not contained in a navigation controller."
                                     userInfo:nil];
    }

    AWXCardViewController *controller = [[AWXCardViewController alloc] initWithNibName:nil bundle:nil];
    controller.session = self.session;

    NSMutableArray<AWXCardScheme *> *supportedCardSchemes = [NSMutableArray array];
    for (AWXCardBrand brand in cardSchemes) {
        AWXCardScheme *cardScheme = [[AWXCardScheme alloc] init];
        cardScheme.name = brand;
        [supportedCardSchemes addObject:cardScheme];
    }

    AWXCardViewModel *viewModel = [[AWXCardViewModel alloc] initWithSession:self.session supportedCardSchemes:supportedCardSchemes launchDirectly:YES];
    controller.viewModel = viewModel;

    [navigationController pushViewController:controller animated:YES];
}

- (void)presentCardPaymentFlowFrom:(UIViewController *)hostViewController {
    [self presentCardPaymentFlowFrom:hostViewController cardSchemes:@[AWXCardBrandVisa, AWXCardBrandMastercard, AWXCardBrandAmex, AWXCardBrandDiscover, AWXCardBrandDinersClub, AWXCardBrandJCB, AWXCardBrandUnionPay]];
}

- (void)pushCardPaymentFlowFrom:(UIViewController *)hostViewController {
    [self pushCardPaymentFlowFrom:hostViewController cardSchemes:@[AWXCardBrandVisa, AWXCardBrandMastercard, AWXCardBrandAmex, AWXCardBrandDiscover, AWXCardBrandDinersClub, AWXCardBrandJCB, AWXCardBrandUnionPay]];
}

@end

File: Airwallex/Card/AWXUIContext+Card.swift
//
//  AWXUIContext+Card.swift
//  Card
//
//  Created by Hector.Huang on 2024/8/23.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import UIKit

public extension AWXUIContext {
    /**
     Present the card payment flow.
     */
    @objc func presentCardPaymentFlow(
        from hostViewController: UIViewController,
        cardSchemes: [AWXCardBrand]
    ) {
        let controller = AWXCardViewController(nibName: nil, bundle: nil)
        controller.session = session
        controller.viewModel = AWXCardViewModel(
            session: session,
            supportedCardSchemes: cardSchemes.map { let cardScheme = AWXCardScheme(); cardScheme.name = $0.rawValue; return cardScheme },
            launchDirectly: true
        )
        
        let navigationController = UINavigationController(rootViewController: controller)
        navigationController.isModalInPresentation = true
        hostViewController.present(navigationController, animated: true)
    }
    
    /**
     Push the card payment flow.
     */
    @objc func pushCardPaymentFlow(
        from hostViewController: UIViewController,
        cardSchemes: [AWXCardBrand]
    ) {
        let navigationController = hostViewController as? UINavigationController ?? hostViewController.navigationController
        guard let navigationController else { 
            fatalError("The hostViewController is not a navigation controller, or is not contained in a navigation controller.")
        }
        
        let controller = AWXCardViewController(nibName: nil, bundle: nil)
        controller.session = session
        controller.viewModel = AWXCardViewModel(
            session: session,
            supportedCardSchemes: cardSchemes.map { let cardScheme = AWXCardScheme(); cardScheme.name = $0.rawValue; return cardScheme },
            launchDirectly: true
        )
        
        navigationController.pushViewController(controller, animated: true)
    }
    
    @objc func presentCardPaymentFlow(from hostViewController: UIViewController) {
        presentCardPaymentFlow(from: hostViewController, cardSchemes: [.visa, .mastercard, .amex, .discover, .dinersClub, .JCB, .unionPay])
    }
    
    @objc func pushCardPaymentFlow(from hostViewController: UIViewController) {
        pushCardPaymentFlow(from: hostViewController, cardSchemes: [.visa, .mastercard, .amex, .discover, .dinersClub, .JCB, .unionPay])
    }
}
